src/main/scala/SerialAdapter.scala

```diff
@@ -208,38 +226,117 @@ class SimSerial(w: Int) extends BlackBox with HasBlackBoxResource {
-    domain.clockNode := tlbus.fixedClockNode
-
-    val inner_io = domain {
-      val adapter = LazyModule(new SerialAdapter)
-      tlbus.fromPort(Some(portName))() := adapter.node
-      InModuleBody {
-        val inner_io = IO(new SerialIO(SERIAL_IF_WIDTH)).suggestName("serial")
-        inner_io.out <> Queue(adapter.module.io.serial.out)
-        adapter.module.io.serial.in <> Queue(inner_io.in)
-        inner_io
-      }
+
+    // TODO: currently assume manager and client buses have same clock
+    domain.clockNode := manager.fixedClockNode
+    val serdesser = domain { LazyModule(new TLSerdesser(
+      w = params.width,
+      clientPortParams = clientPortParams,
+      managerPortParams = managerPortParams
+    )) }
+    manager.coupleTo(s"port_named_serial_tl_mem") {
+      (serdesser.managerNode
+        := TLBuffer()
+        := TLSourceShrinker(1 << memParams.idBits)
+        := TLWidthWidget(manager.beatBytes)
+        := _ )
+    }
+    client.coupleFrom(s"port_named_serial_tl_ctrl") {
+      ( _
+        := TLBuffer(BufferParams.default)
+        := serdesser.clientNode
+      )
     }
+
+    val inner_io = domain { InModuleBody {
+      val inner_io = IO(new SerialIO(params.width)).suggestName("serial_tl")
+      inner_io.out <> serdesser.module.io.ser.out
+      serdesser.module.io.ser.in <> inner_io.in
+      inner_io
+    } }
     val outer_io = InModuleBody {
-      val outer_io = IO(new ClockedIO(new SerialIO(SERIAL_IF_WIDTH))).suggestName("serial")
+      val outer_io = IO(new ClockedIO(new SerialIO(params.width))).suggestName("serial_tl")
       outer_io.bits <> inner_io
       outer_io.clock := domain.module.clock
       outer_io
     }
-    Some(outer_io)
-  } else {
-    None
+    (Some(serdesser), Some(outer_io))
+  }.getOrElse(None, None)
+}
+
+class SerialRAM(
+  w: Int,
+  memParams: MasterPortParams,
+  managerEdge: TLEdgeParameters,
+  clientEdge: TLEdgeParameters)(implicit p: Parameters) extends LazyModule {
+  val managerParams = clientEdge.slave // the managerParams are the chip-side clientParams
+  val clientParams = managerEdge.master // The clientParams are the chip-side managerParams
+  val adapter = LazyModule(new SerialAdapter)
+  val serdesser = LazyModule(new TLSerdesser(
+    w,
+    clientParams,
+    managerParams
+  ))
+
+  val beatBytes = memParams.beatBytes
+  val base = memParams.base
+  val size = memParams.size
+  val srams = AddressSet.misaligned(base, size).map { aset =>
+    LazyModule(new TLRAM(
+      aset,
+      beatBytes = beatBytes
+    ))
+  }
+  val xbar = TLXbar()
+  srams.foreach { s => s.node := TLBuffer() := TLFragmenter(beatBytes, p(CacheBlockBytes)) := xbar }
+  xbar := serdesser.clientNode
+
+  serdesser.managerNode := TLBuffer() := adapter.node
+
+  lazy val module = new LazyModuleImp(this) {
+    val io = IO(new Bundle {
+      val ser = Flipped(new SerialIO(w))
+      val tsi_ser = new SerialIO(SERIAL_TSI_WIDTH)
```

Comment:
If the adapter is inside of this module do we need to punch this out?

Reply:
Ah I forgot that SerialAdapter doesn't instantiate the SimSerial component. Ignore.